[train] Fix the xgboost v2 callback

liulehui opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

1. in the ray train [revamp REP](https://github.com/ray-project/enhancements/blob/main/reps/2024-10-18-train-tune-api-revamp/2024-10-18-train-tune-api-revamp.md#tune-only-usage), we decouple the ray train/ray tune dependency.
2. Hence, when using RayTrainReportCallback when reporting metrics or checkpoint: e.g. in this [line](https://github.com/ray-project/ray/blob/master/python/ray/train/xgboost/_xgboost_utils.py#L170), the v2 context api will throw [RuntimeError](https://github.com/ray-project/ray/blob/master/python/ray/train/v2/_internal/execution/context.py#L279-L283).
3. In V1 this issue is mitigated by [switch to Tune Context](https://github.com/ray-project/ray/blob/master/python/ray/train/context.py#L126-L128) when train.get_context() is called.
4. In order to make the xgboost tune only usage callback continue working, hence the bypass the use `_is_tune_session()` to get context for this callback explicitly if this is used in tune only when we trying to get train context in V2 manner and resolve ray.tune.report is tune only based on migration example [here](https://github.com/ray-project/enhancements/blob/main/reps/2024-10-18-train-tune-api-revamp/2024-10-18-train-tune-api-revamp.md#tune-only-usage).

## Related issue number

NA

## Checks

- [x ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] tested locally 
   - [ ] Release tests
   - [ ] This PR is not tested :(

liulehui committed:
Fix the xgboost v2 callback

Signed-off-by: Lehui Liu <lehui@anyscale.com>

liulehui committed:
fix the xgboost v2 callback

Signed-off-by: Lehui Liu <lehui@anyscale.com>

liulehui commented on python/ray/train/xgboost/_xgboost_utils.py:
done aded PR description. PTAL.

matthewdeng commented on python/ray/train/xgboost/_xgboost_utils.py:
```diff
@@ -167,7 +171,7 @@ def _get_report_dict(self, evals_log):
     @contextmanager
     def _get_checkpoint(self, model: Booster) -> Optional[Checkpoint]:
         # NOTE: The world rank returns None for Tune usage without Train.
-        if ray.train.get_context().get_world_rank() in (0, None):
+        if _in_tune_session() or ray.train.get_context().get_world_rank() in (0, None):
```
Is this note still applicable? In which case it would be okay to not include the new condition.

liulehui replied:
I think this note is still applicable in v1 Tune without train, but this case can be caught by the newly added `_in_tune_session()`.
But how about we remove the note and change to  
`if _in_tune_session() or ray.train.get_context().get_world_rank() == 0` ?

justinvyu replied:
Since this xgboost callback can be passed into xgboost.train regardless of whether you're running with Train or Tune, this method needs to handle both the distributed case and the non-distributed case.

We have an Tune alias of this callback: `ray.tune.integration.xgboost.TuneReportCheckpointCallback`, which is what we recommend in [Tune user guides](https://docs.ray.io/en/latest/tune/examples/tune-xgboost.html).
Running this Tune callback with the RAY_TRAIN_V2_ENABLED=1 flag was the original issue the user flagged.


What about making `ray.tune.integration.xgboost.TuneReportCheckpointCallback` a real subclass which implements `report_fn` and `get_checkpoint` differently from the Train version?

matthewdeng replied:
Ya I think this would be the most clean, can also create a single abstract class that both the Train and Tune callbacks implement.

matthewdeng commented on python/ray/train/xgboost/_xgboost_utils.py:
Can you double check that this change works as expected for these scenarios?

| RAY_TRAIN_V2_ENABLED| 0 | 1 |
|----|--------|--------|
| Train |  Yes | No |
| Tune | Yes |  Yes |

liulehui replied:
Sure, tested with the below result:
| RAY_TRAIN_V2_ENABLED| 0 | 1 |
|----|--------|--------|
| Train |  Yest | Failed with `ValueError: All callbacks must be instances of `ray.train.UserCallback`. Passing in a Ray Tune callback is no longer supported.|
| Tune | Yes |  Yes |

justinvyu replied:
This callback is meant to be passed into `xgboost.train(callbacks)`, not the `Tuner/TorchTrainer(run_config)` callbacks. Can we try again passing it into xgboost?

liulehui committed:
Merge branch 'master' into fix-v2-callback

liulehui committed:
Add a abstract class RayReportCallback

1. make `ray.train.xgboost.RayTrainReportCallback` and `ray.tune.integration.xgboost.TuneReportCheckpointCall` implement it
2. each subclass define its own report_fn and get_checkpoint method
3. common logic stays in the abstract class

Signed-off-by: Lehui Liu <lehui@anyscale.com>

liulehui committed:
resolve comment

Signed-off-by: Lehui Liu <lehui@anyscale.com>

liulehui commented on python/ray/train/xgboost/_xgboost_utils.py:
```diff
@@ -164,15 +99,15 @@ def _get_report_dict(self, evals_log):
+    @abstractmethod
+    def _get_checkpoint(
+        self, model: Booster
+    ) -> Union[Optional[Checkpoint], Optional[RayTuneCheckpoint]]:
```
done.

liulehui commented on python/ray/train/xgboost/_xgboost_utils.py:
```diff
@@ -19,83 +21,11 @@ class TrainingCallback:
+class RayReportCallback(TrainingCallback):
     CHECKPOINT_NAME = "model.ubj"
 
+    report_fn: Callable[[Dict, Optional[Checkpoint]], None] = ray.train.report
```
I separate out the report part, this way, if later we need to further differentiate train and tune, we can implement in its corresponding report functions.

liulehui committed:
remove unused report_fn

Signed-off-by: Lehui Liu <lehui@anyscale.com>

liulehui commented on python/ray/train/xgboost/_xgboost_utils.py:
```diff
@@ -124,7 +54,9 @@ def __init__(
-        cls, checkpoint: Checkpoint, filename: str = CHECKPOINT_NAME
+        cls,
+        checkpoint: Union[Checkpoint, RayTuneCheckpoint],
+        filename: str = CHECKPOINT_NAME,
```
removed.

liulehui committed:
Merge branch 'master' into fix-v2-callback

liulehui committed:
Merge branch 'ray-project:master' into fix-v2-callback

liulehui committed:
revert back the lint file

Signed-off-by: Lehui Liu <lehui@anyscale.com>

liulehui commented on python/ray/train/xgboost/_xgboost_utils.py:
```diff
@@ -19,22 +20,101 @@ class TrainingCallback:
+        pass
+
+    @abstractmethod
     def after_iteration(self, model: Booster, epoch: int, evals_log: Dict):
-        raise NotImplementedError
+        pass
+
+    @abstractmethod
+    def after_training(self, model: Booster) -> Booster:
+        pass
```
done! Thanks for the pointer!

liulehui committed:
resolve comments

Signed-off-by: Lehui Liu <lehui@anyscale.com>

liulehui committed:
add back the pydoclint

Signed-off-by: Lehui Liu <lehui@anyscale.com>

liulehui commented on ci/lint/pydoclint-baseline.txt:
added back

matthewdeng commented on python/ray/train/xgboost/_xgboost_utils.py:
```diff
@@ -164,15 +96,29 @@ def _get_report_dict(self, evals_log):
+
+        This method needs to be implemented by subclasses.
+        """
+        pass
```
Let's `raise NotImplementedError` instead for these abstract methods.

liulehui replied:
done!

matthewdeng commented on python/ray/train/xgboost/_xgboost_utils.py:
```diff
@@ -186,10 +132,10 @@ def after_iteration(self, model: Booster, epoch: int, evals_log: Dict):
-            with self._get_checkpoint(model=model) as checkpoint:
-                ray.train.report(report_dict, checkpoint=checkpoint)
+            self._save_and_report_checkpoint(report_dict, model)
+
         else:
-            ray.train.report(report_dict)
+            self._report_metrics(report_dict)
```
so clean!

liulehui committed:
raise NotImplementedError

Signed-off-by: Lehui Liu <lehui@anyscale.com>

justinvyu reviewed (approved):
Thanks for the fix! 🚀

liulehui committed:
Merge branch 'master' into fix-v2-callback

liulehui committed:
Merge branch 'master' into fix-v2-callback

liulehui committed:
Merge branch 'master' into fix-v2-callback

liulehui committed:
Merge branch 'master' into fix-v2-callback

liulehui committed:
Merge branch 'master' into fix-v2-callback

liulehui committed:
try to fix pydoclint

Signed-off-by: Lehui Liu <lehui@anyscale.com>

liulehui committed:
Merge branch 'master' into fix-v2-callback

matthewdeng commented on ci/lint/pydoclint-baseline.txt:
```diff
@@ -125,9 +125,6 @@ python/ray/_private/resource_isolation_config.py
     DOC103: Method `ResourceIsolationConfig.__init__`: Docstring arguments are different from function arguments. (Or could be other formatting issues: https://jsh9.github.io/pydoclint/violation_codes.html#notes-on-doc103 ). Arguments in the function signature but not in the docstring: [cgroup_path: Optional[str], enable_resource_isolation: bool, system_reserved_cpu: Optional[float], system_reser ...
     DOC201: Method `ResourceIsolationConfig._validate_and_get_system_reserved_cpu` does not have a return section in docstring
 --------------------
-python/ray/_private/resource_spec.py
-    DOC201: Method `ResourceSpec.resolve` does not have a return section in docstring
---------------------
```
Is this removal expected?
